src/cascadia/TerminalCore/Terminal.cpp

```diff
@@ -1348,6 +1345,97 @@ void Terminal::_updateUrlDetection()
     }
 }
 
+// Interns URegularExpression instances so that they can be reused. This method is thread-safe.
+static ICU::unique_uregex internURegularExpression(const std::wstring_view& pattern)
+{
+    struct CacheValue
+    {
+        ICU::unique_uregex re;
+        size_t generation = 0;
+    };
+
+    struct CacheKeyHasher
+    {
+        using is_transparent = void;
+
+        std::size_t operator()(const std::wstring_view& str) const
+        {
+            return til::hash(str);
+        }
+    };
+
+    struct SharedState
+    {
+        wil::srwlock lock;
+        std::unordered_map<std::wstring, CacheValue, CacheKeyHasher, std::equal_to<>> set;
+        size_t totalInsertions = 0;
+    };
+
+    static SharedState shared;
+    static constexpr size_t cacheSizeLimit = 128;
+
+    UErrorCode status = U_ZERO_ERROR;
+
+    {
+        const auto guard = shared.lock.lock_shared();
+        if (const auto it = shared.set.find(pattern); it != shared.set.end())
+        {
+            return ICU::unique_uregex{ uregex_clone(it->second.re.get(), &status) };

```

Comment:
Returned objects are always being moved, but with much stronger guarantees than regular `std::move`. `return` will behave as if there was no function call at all. In other words, returning a `std::string` by value is like constructing that string in the caller directly. 0 overhead.
`std::move` is basically just a `T&` with extra syntactic sugar. So, if you `return std::move(string);` it will actually be forced to call the `string(string&&)` constructor on the way out, because you're not returning an object value anymore - you're returning a reference which now needs to be converted to a value. That's why one should never `return std::move`.
The reason it can be seen anyways sometimes is because when you have code like this:
```cpp
struct StringWrapper {
    // implicit constructor!
    StringWrapper(std::string&& str) {}
};

StringWrapper foobar() {
    std::string str = "...";
    // calls the implicit constructor
    return std::move(str);
}
```

`getFontCollection()` in `FontCache.h` is an example for that.